sandbox/grist/formula_prompt.py

```diff
@@ -107,21 +110,50 @@ def visible_columns(engine, table_id):
   ]
 
 
-def class_schema(engine, table_id, exclude_col_id=None):
+def class_schema(engine, table_id, exclude_col_id=None, lookups=False):
   result = "@dataclass\nclass {}:\n".format(table_id)
+
+  if lookups:
+    tmp = []
+    tmp2 = []
+    for col_id, col in visible_columns(engine, table_id):
+      if col_id != exclude_col_id:
+        tmp.append(col_id + ' = None')
+        tmp2.append(f'{col_id}={col_id}')
+    tmp.append('sort_by = None')
+    tmp2.append('sort_by=sort_by')
+    args = ', '.join(tmp)
+    args2 = ', '.join(tmp2)
+    result += f"     def __len__(self):\n"
+    result += f"        return len({table_id}.lookupRecords())\n"
+    result += f"    @staticmethod\n"
+    result += f"    def lookupRecords({args}) -> List[{table_id}]:\n"
+    result += f"       # ...\n"
```

Comment:
Suggestion:

```python
class _Table:
    @classmethod
    def lookupRecords(cls: Type[_T], **kwargs, sort_by=None) -> List[_T]:
        """
        Returns records matching the given keyword arguments.
        To control order, use e.g. `sort_by='Key' or `sort_by='-Key'`.
        """
        result = [r for r in cls.all if all(getattr(r, k) == v for k, v in kwargs.items())]
        if sort_by:
            result.sort(key=lambda r: getattr(r, sort_by.lstrip("-")), reverse=sort_by.startswith("-"))
        return result


@dataclasses
class Table1(_Table):
```

Haven't experimented directly, but copilot filled a lot of this in by itself, so I think it understands it.

Reply:
Thanks, will bear it in mind. So far the minimal prompting has been adequate, and I feel a bit bad about the prompt bloat I'm adding so I'm trying to keep it short (for now).

Reply:
Sure, the idea is that it would make the prompt shorter when there's many tables by not defining the same three methods on each one, but I get how that comes with caveats.